### PR TITLE
#8889: standalone mod component activation adapter

### LINKED ISSUE #8888: Slice 1: Implement the cleanup logic in getActivatedModComponentFromDefinition to handle standalone mod components
https://www.notion.so/pixiebrix/Extension-Console-Mod-Activation-remove-standalone-mod-activation-b29abc71d47e44b7ab36b4a55ddfa0f6?pvs=4#4dcee695aece46ea803c2ff900b00ee5

### LINKED ISSUE #8889: Slice 2: Refactor ActivateExtensionPage to transform the standalone mod component into a mod, and then pass it into ActivateModCard
https://www.notion.so/pixiebrix/Extension-Console-Mod-Activation-remove-standalone-mod-activation-b29abc71d47e44b7ab36b4a55ddfa0f6?pvs=4#b22324bb65ac437c957010a341950e2c

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Closes #8888
- Closes #8889
- Part of #8890 (see discussion, future work)

## Reviewer Notes

- In the spec/sketch we missed that we need to ensure the activated mod component has the same id as the id on the server. Otherwise, the mods screen shows an entry for the activated mod, and the standalone definition on the server

# Remaining Work

- [x] Fix bug where mod shown 2x on mod screen after activation because the activated mod component has a different id than the mod component on the server

## Known Bugs

- 🐛 When you activate a standalone mod component, the update timestamp on the mods screen shows as the time the mod was activated, not the update timestamp on the definition. That's because of the [`getUpdatedAt`](https://github.com/pixiebrix/pixiebrix-extension/blob/7a98bd07ad2047e0b58b6b104d6d6188bd473111/src/utils/modUtils.ts#L238-L238) method uses the mod activation timestamp, not the definition on the server. This bug has a negligible impact, and will likely be solved by the next phase, so I recommend we don't address it now. The potential fix would be to change `forceModComponentId` to also override the timestamp. But that complicates the async state of the page

## Future Work

- Playwright test: do we have a way to get standalone mod component fixtures?
- 😞 Can't remove `extensionsSlice.activateStandaloneModDefinition` because it's used in test setup for `ConvertToModModal` and `useModPageActions`
- 😞 Strict null checks: `useActivateModWizard` is not null-safe yet, so can't add files to strict null checks
- Remove `forceModComponentId` code once Mod Screen / extensionsSlice state is updated to use internal mod ids for standalone mod components
- Make `modUtils` less confusing. It's confusing that all of the methods check for `_recipe` on the `isStandaloneModComponent` branch given that on the Mods Screen that branch is only used for standalone mod components. (From the types alone though, it is allowable for a mod component from a mod to be passed to the method). Should naturally be addressed by the next phase: https://www.notion.so/pixiebrix/Extension-Console-Mod-Screen-shift-left-conversion-to-common-mod-shape-d457621fc83d495bbeb135cbd3977ee6?pvs=4

For more information on our expectations for the PR process, see the
[code review principles doc](https://www.notion.so/pixiebrix/Code-Review-Principles-1ce7276b82a84d2a995d55ad85e1310d?pvs=4)

### CODE REVIEW

File: src/activation/mapModComponentDefinitionToActivatedModComponent.ts [hunk 1]
```diff
@@ -40,13 +41,13 @@ export type ActivateModComponentParam = {
...
  * @param modComponentDefinition the component definition to activate
- * @param apiVersion the pixiebrix mod api version
+ * @param apiVersion the PixieBrix definition API version
  * @param modDefinition the mod definition this component belongs to
  * @param deployment the deployment for this component, if it belongs to one
  * @param optionsArgs mod option inputs for the mod this component belongs to
  * @param integrationDependencies the configured dependencies for the mod this component belongs to
  */
-export function getActivatedModComponentFromDefinition<
+export function mapModComponentDefinitionToActivatedModComponent<
```
Comment (twschiller) [inline src/activation/mapModComponentDefinitionToActivatedModComponent.ts:50]: Follow standard `mapSourceToDestination` convention

File: src/extensionConsole/pages/activateMod/ActivateModDefinitionIdPage.test.tsx
Comment (twschiller) [inline src/extensionConsole/pages/activateMod/ActivateModDefinitionIdPage.test.tsx:1]: This file was testing the Page, not the Card. So fixed the file name

File: src/types/helpers.ts [hunk 2]
```diff
@@ -127,30 +123,6 @@ export function validateRegistryId(id: string | undefined): RegistryId {
   throw new Error("Invalid registry id");
 }
 
-function isTimestamp(value: string): value is Timestamp {
```
Comment (twschiller) [inline src/types/helpers.ts:130]: Moved to timeUtils

File: src/utils/timeUtils.ts [hunk 3]
```diff
@@ -16,6 +16,38 @@
...
+import { type Timestamp } from "@/types/stringTypes";
+
+/**
+ * Returns the current time in ISO format.
+ */
+export function isoTimestamp(): Timestamp {
+  return new Date().toISOString() as Timestamp;
+}
+
+function isTimestamp(value: string): value is Timestamp {
```
Comment (twschiller) [inline src/utils/timeUtils.ts:28]: Moved from helpers file

File: src/activation/useActivateMod.ts [hunk 4]
```diff
@@ -179,7 +190,13 @@ function useActivateMod(
         success: true,
       };
     },
-    [createDatabase, dispatch, activatedModComponents, source],
+    [
+      createDatabase,
+      dispatch,
+      activatedModComponents,
+      source,
+      checkPermissions,
```
Comment (twschiller) [inline src/activation/useActivateMod.ts:198]: Add missing dependency

File: src/extensionConsole/pages/activateMod/ActivateModCard.tsx [hunk 5]
```diff
@@ -61,7 +56,9 @@ const WizardHeader: React.VoidFunctionComponent<{
         </span>
         <span>
           <Card.Title>{mod.metadata.name}</Card.Title>
-          <code className={styles.packageId}>{mod.metadata.id}</code>
+          {!isInternalRegistryId(mod.metadata.id) && (
```
Comment (twschiller) [inline src/extensionConsole/pages/activateMod/ActivateModCard.tsx:59]: Prevent showing internal id on the activation screen

File: src/extensionConsole/pages/activateMod/ActivateModCard.tsx [hunk 6]
```diff
@@ -75,34 +72,25 @@ const WizardHeader: React.VoidFunctionComponent<{
...
-  } = useActivateModWizard(mod);
+  } = useActivateModWizard(modDefinition);
 
   const activateMod = useActivateMod("extensionConsole");
   const [activationError, setActivationError] = useState<unknown>();
   const [createMilestone] = useCreateMilestoneMutation();
 
   const { hasMilestone } = useMilestones();
 
-  useSetDocumentTitle(
```
Comment (twschiller) [inline src/extensionConsole/pages/activateMod/ActivateModCard.tsx:100]: Move tot he parent component

File: src/modDefinitions/util/pickModDefinitionMetadata.ts [hunk 7]
```diff
@@ -30,6 +34,10 @@ export function pickModDefinitionMetadata(
     throw new TypeError("ModDefinition metadata id is required");
   }
 
+  if (isInternalRegistryId(modDefinition.metadata.id)) {
```
Comment (twschiller) [inline src/modDefinitions/util/pickModDefinitionMetadata.ts:37]: Here's where the metadata gets dropped. In future phases we might decide it's better to be located in the caller

File: src/types/helpers.ts [hunk 8]
```diff
@@ -15,7 +15,7 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-import { valid as semVerValid, coerce as semVerCoerce } from "semver";
+import { coerce as semVerCoerce, valid as semVerValid } from "semver";
```
Comment (twschiller) [inline src/types/helpers.ts:18]: IntelliJ just organizing imports

File: src/data/service/api.ts [hunk 9]
```diff
@@ -236,14 +236,14 @@ export const appApi = createApi({
     }),
     getStandaloneModDefinition: builder.query<
       StandaloneModDefinition,
-      { extensionId: UUID }
+      { modComponentId: UUID }
```
Comment (twschiller) [inline src/data/service/api.ts:239]: 💅

File: src/extensionConsole/App.tsx [hunk 10]
```diff
@@ -32,8 +32,8 @@ import Sidebar from "@/extensionConsole/Sidebar";
 import { Route, Switch } from "react-router-dom";
 import { ConnectedRouter } from "connected-react-router";
 import EnvironmentBanner from "@/layout/EnvironmentBanner";
-import ActivateModPage from "@/extensionConsole/pages/activateMod/ActivateModPage";
-import ActivateExtensionPage from "@/extensionConsole/pages/activateExtension/ActivateExtensionPage";
+import ActivateModDefinitionIdPage from "@/extensionConsole/pages/activateMod/ActivateModDefinitionIdPage";
+import ActivateStandaloneModDefinitionIdPage from "@/extensionConsole/pages/activateMod/ActivateStandaloneModDefinitionIdPage";
```
Comment (BLoe) [inline src/extensionConsole/App.tsx:36]: Was it intentional to put `Id` in the names of these two components? I think it makes the names a little bit confusing.

Comment (twschiller) [reply src/extensionConsole/App.tsx:36]: Yes - these are the wrapper components that then pass the mod definition down. I'm open to other component names

Comment (BLoe) [reply src/extensionConsole/App.tsx:36]: Ok, I'm seeing the other components now, never mind

**Review by BLoe — APPROVED** [review]

**Pull request merged**